rig/machine_control/machine_controller.py

```diff
@@ -1739,34 +1807,169 @@ def get_machine(self, x=0, y=0, default_num_cores=18):
         chip_resource_exceptions = {}
 
         # Discover dead links and cores
-        for (x, y), p2p_route in iteritems(p2p_tables):
-            if x <= max_x and y <= max_y:
-                if p2p_route == consts.P2PTableEntry.none:
+        for x in range(width):
+            for y in range(height):
+                chip_info = chips_info.get((x, y))
+                if chip_info is None:
                     dead_chips.add((x, y))
                 else:
-                    try:
-                        num_working_cores = self.get_num_working_cores(x, y)
-                        working_links = self.get_working_links(x, y)
-
-                        if num_working_cores < default_num_cores:
-                            resource_exception = chip_resources.copy()
-                            resource_exception[Cores] = min(default_num_cores,
-                                                            num_working_cores)
-                            chip_resource_exceptions[(x, y)] = \
-                                resource_exception
-
-                        for link in set(Links) - working_links:
-                            dead_links.add((x, y, link))
-                    except SCPError:
-                        # The chip was listed in the P2P table but is not
-                        # responding. Assume it is dead anyway.
-                        dead_chips.add((x, y))
+                    if chip_info.num_cores < default_num_cores:
+                        resource_exception = chip_resources.copy()
+                        resource_exception[Cores] = min(default_num_cores,
+                                                        chip_info.num_cores)
```

Comment:
You don't; it may be equal; `chip_info.num_cores` is the number working of cores present (which is usually 17 or 18), not just the number which are idle (which generally would be strictly < 18).